File: ulib/js/FStar_CommonST.js
/* @flow */

export let read = <T>(x:T): T => x;

export let op_Bang = <T>(x:T): T => read(x);

export let write = <T>(x:T): ((y:T) => null) => ((y:T) => {x = y; return null});

export let op_Colon_Equals = <T>(x:T): ((y:T) => null) => ((y:T) => write(x)(y));

export let uid = 0;

export let alloc = <T>(contents:T): T => contents;

export let recall: (<T>(r:T) => null) = (r) => null;

export let get = () => null;

File: stage0/ulib/fs/FStar_CommonST.fs
module FStar_CommonST

open FStar_Monotonic_Heap

let read x = x.contents

let op_Bang x = read x

let write x y = x.contents <- y

let op_Colon_Equals x y = write x y

let uid = ref 0

let alloc contents =
  let id = incr uid; !uid in
  let r = { id = id; contents = contents } in
  r

let recall = (fun r -> ())
let get () = ()

type 'a witnessed = | C

let gst_witness = (fun r -> ())
let gst_recall = (fun r -> ())
